File: governance/engine/orchestrator/circuit_breaker.py
"""Per-work-unit evaluation cycle tracking with hard limits.

Enforces the circuit breaker rules from agent-protocol.md:
- Max 2 Tester FEEDBACK cycles per work unit (trips on 3rd — Tester
  must emit BLOCK at cycle 3 per protocol)
- Max 5 total evaluation cycles per work unit (includes re-assigns after BLOCK/ESCALATE)
"""

from __future__ import annotations

from dataclasses import dataclass, field


class CircuitBreakerTripped(Exception):
    """Raised when a work unit exceeds its evaluation cycle limit."""

    def __init__(self, correlation_id: str, reason: str, cycles: int, limit: int):
        self.correlation_id = correlation_id
        self.reason = reason
        self.cycles = cycles
        self.limit = limit
        super().__init__(
            f"Circuit breaker tripped for {correlation_id}: "
            f"{reason} ({cycles}/{limit})"
        )


@dataclass
class WorkUnit:
    """Tracks evaluation cycles for a single work item (issue/PR)."""

    correlation_id: str
    feedback_cycles: int = 0
    total_eval_cycles: int = 0
    blocked: bool = False
    block_reason: str | None = None


class CircuitBreaker:
    """Tracks per-work-unit evaluation cycles and enforces hard limits.

    Constants from agent-protocol.md:
        MAX_FEEDBACK_CYCLES = 2 (Tester FEEDBACK only — trips on 3rd per protocol)
        MAX_TOTAL_EVAL_CYCLES = 5 (FEEDBACK + re-ASSIGN after BLOCK/ESCALATE)
    """

    def __init__(
        self,
        max_feedback_cycles: int = 2,
        max_total_eval_cycles: int = 5,
    ):
        self.max_feedback_cycles = max_feedback_cycles
        self.max_total_eval_cycles = max_total_eval_cycles
        self._work_units: dict[str, WorkUnit] = {}

    def _get_or_create(self, correlation_id: str) -> WorkUnit:
        if correlation_id not in self._work_units:
            self._work_units[correlation_id] = WorkUnit(
                correlation_id=correlation_id
            )
        return self._work_units[correlation_id]

    def record_feedback(self, correlation_id: str) -> WorkUnit:
        """Record a Tester FEEDBACK cycle.

        Raises:
            CircuitBreakerTripped: If feedback cycle limit (2, trips on 3rd)
                or total eval cycle limit (5) is exceeded.
        """
        unit = self._get_or_create(correlation_id)
        if unit.blocked:
            raise CircuitBreakerTripped(
                correlation_id, "work_unit_already_blocked",
                unit.total_eval_cycles, self.max_total_eval_cycles,
            )

        unit.feedback_cycles += 1
        unit.total_eval_cycles += 1

        if unit.feedback_cycles > self.max_feedback_cycles:
            unit.blocked = True
            unit.block_reason = "feedback_cycle_limit"
            raise CircuitBreakerTripped(
                correlation_id, "feedback_cycle_limit",
                unit.feedback_cycles, self.max_feedback_cycles,
            )

        if unit.total_eval_cycles >= self.max_total_eval_cycles:
            unit.blocked = True
            unit.block_reason = "total_eval_cycle_limit"
            raise CircuitBreakerTripped(
                correlation_id, "total_eval_cycle_limit",
                unit.total_eval_cycles, self.max_total_eval_cycles,
            )

        return unit

    def record_reassign(self, correlation_id: str) -> WorkUnit:
        """Record a re-ASSIGN after BLOCK or ESCALATE.

        Raises:
            CircuitBreakerTripped: If total eval cycle limit (5) is exceeded.
        """
        unit = self._get_or_create(correlation_id)
        if unit.blocked:
            raise CircuitBreakerTripped(
                correlation_id, "work_unit_already_blocked",
                unit.total_eval_cycles, self.max_total_eval_cycles,
            )

        unit.total_eval_cycles += 1

        if unit.total_eval_cycles >= self.max_total_eval_cycles:
            unit.blocked = True
            unit.block_reason = "total_eval_cycle_limit"
            raise CircuitBreakerTripped(
                correlation_id, "total_eval_cycle_limit",
                unit.total_eval_cycles, self.max_total_eval_cycles,
            )

        return unit

    def can_dispatch(self, correlation_id: str) -> bool:
        """Check if a work unit can accept more evaluation cycles."""
        unit = self._work_units.get(correlation_id)
        if unit is None:
            return True
        return not unit.blocked

    def get_unit(self, correlation_id: str) -> WorkUnit | None:
        """Get work unit state. Returns None if not tracked."""
        return self._work_units.get(correlation_id)

    @property
    def all_units(self) -> dict[str, WorkUnit]:
        """Read-only view of all tracked work units."""
        return dict(self._work_units)

File: src/internal/orchestrator/circuitbreaker/circuitbreaker.go
// Package circuitbreaker implements per-work-unit failure detection for
// the orchestrator. When a work unit exceeds its feedback or evaluation
// cycle limits the circuit trips, preventing further dispatch.
//
// Ported from Python: governance/engine/orchestrator/circuit_breaker.py
package circuitbreaker

import "fmt"

// ---------------------------------------------------------------------------
// WorkUnit
// ---------------------------------------------------------------------------

// WorkUnit tracks failure counters for a single work item (issue / PR).
type WorkUnit struct {
	ID                 string `json:"id"`
	FeedbackCycles     int    `json:"feedback_cycles"`
	TotalEvalCycles    int    `json:"total_eval_cycles"`
	Reassignments      int    `json:"reassignments"`
	MaxFeedbackCycles  int    `json:"max_feedback_cycles"`
	MaxTotalEvalCycles int    `json:"max_total_eval_cycles"`
	Tripped            bool   `json:"tripped"`
}

// ---------------------------------------------------------------------------
// CircuitBreaker
// ---------------------------------------------------------------------------

// CircuitBreaker manages a collection of WorkUnits.
type CircuitBreaker struct {
	units              map[string]*WorkUnit
	maxFeedbackCycles  int
	maxTotalEvalCycles int
}

// New creates a CircuitBreaker. Defaults: maxFeedback=2, maxTotal=5.
func New(maxFeedback, maxTotal int) *CircuitBreaker {
	if maxFeedback <= 0 {
		maxFeedback = 2
	}
	if maxTotal <= 0 {
		maxTotal = 5
	}
	return &CircuitBreaker{
		units:              make(map[string]*WorkUnit),
		maxFeedbackCycles:  maxFeedback,
		maxTotalEvalCycles: maxTotal,
	}
}

// ensureUnit lazily initializes a WorkUnit.
func (cb *CircuitBreaker) ensureUnit(workID string) *WorkUnit {
	if u, ok := cb.units[workID]; ok {
		return u
	}
	u := &WorkUnit{
		ID:                 workID,
		MaxFeedbackCycles:  cb.maxFeedbackCycles,
		MaxTotalEvalCycles: cb.maxTotalEvalCycles,
	}
	cb.units[workID] = u
	return u
}

// RecordFeedback increments the feedback cycle counter for workID.
// Returns an error if the circuit trips.
func (cb *CircuitBreaker) RecordFeedback(workID string) error {
	u := cb.ensureUnit(workID)
	u.FeedbackCycles++
	u.TotalEvalCycles++
	if u.FeedbackCycles >= u.MaxFeedbackCycles || u.TotalEvalCycles >= u.MaxTotalEvalCycles {
		u.Tripped = true
		return fmt.Errorf("circuit tripped for %s: feedback=%d/%d total=%d/%d",
			workID, u.FeedbackCycles, u.MaxFeedbackCycles, u.TotalEvalCycles, u.MaxTotalEvalCycles)
	}
	return nil
}

// RecordReassign increments the reassignment counter.
func (cb *CircuitBreaker) RecordReassign(workID string) {
	u := cb.ensureUnit(workID)
	u.Reassignments++
}

// CanDispatch returns true if the work unit has not tripped.
func (cb *CircuitBreaker) CanDispatch(workID string) bool {
	u, ok := cb.units[workID]
	if !ok {
		return true
	}
	return !u.Tripped
}

// GetUnit returns the WorkUnit for workID, or nil if none exists.
func (cb *CircuitBreaker) GetUnit(workID string) *WorkUnit {
	return cb.units[workID]
}

// ToDict serializes the circuit breaker state to a map.
func (cb *CircuitBreaker) ToDict() map[string]interface{} {
	units := make(map[string]interface{}, len(cb.units))
	for id, u := range cb.units {
		units[id] = map[string]interface{}{
			"id":                  u.ID,
			"feedback_cycles":     u.FeedbackCycles,
			"total_eval_cycles":   u.TotalEvalCycles,
			"reassignments":       u.Reassignments,
			"max_feedback_cycles": u.MaxFeedbackCycles,
			"max_total_eval_cycles": u.MaxTotalEvalCycles,
			"tripped":             u.Tripped,
		}
	}
	return map[string]interface{}{
		"units":                units,
		"max_feedback_cycles":  cb.maxFeedbackCycles,
		"max_total_eval_cycles": cb.maxTotalEvalCycles,
	}
}

// FromDict restores a CircuitBreaker from a serialized map.
func FromDict(data map[string]interface{}) *CircuitBreaker {
	cb := &CircuitBreaker{
		units:              make(map[string]*WorkUnit),
		maxFeedbackCycles:  toInt(data["max_feedback_cycles"]),
		maxTotalEvalCycles: toInt(data["max_total_eval_cycles"]),
	}
	if cb.maxFeedbackCycles <= 0 {
		cb.maxFeedbackCycles = 2
	}
	if cb.maxTotalEvalCycles <= 0 {
		cb.maxTotalEvalCycles = 5
	}

	if units, ok := data["units"].(map[string]interface{}); ok {
		for id, raw := range units {
			m, ok := raw.(map[string]interface{})
			if !ok {
				continue
			}
			u := &WorkUnit{
				ID:                 id,
				FeedbackCycles:     toInt(m["feedback_cycles"]),
				TotalEvalCycles:    toInt(m["total_eval_cycles"]),
				Reassignments:      toInt(m["reassignments"]),
				MaxFeedbackCycles:  toInt(m["max_feedback_cycles"]),
				MaxTotalEvalCycles: toInt(m["max_total_eval_cycles"]),
			}
			if tripped, ok := m["tripped"].(bool); ok {
				u.Tripped = tripped
			}
			if u.MaxFeedbackCycles <= 0 {
				u.MaxFeedbackCycles = cb.maxFeedbackCycles
			}
			if u.MaxTotalEvalCycles <= 0 {
				u.MaxTotalEvalCycles = cb.maxTotalEvalCycles
			}
			cb.units[id] = u
		}
	}
	return cb
}

// toInt converts an interface{} to int, handling both int and float64.
func toInt(v interface{}) int {
	switch n := v.(type) {
	case int:
		return n
	case float64:
		return int(n)
	case int64:
		return int(n)
	default:
		return 0
	}
}
